hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -1084,17 +1082,34 @@ def draw_float(
                 result = self._draw_float(
                     forced_sign_bit=forced_sign_bit, forced=forced
                 )
-                if math.copysign(1.0, result) == -1:
-                    assert neg_clamper is not None
-                    clamped = -neg_clamper(-result)
+                # if we drew a nan and we allow nans, great! take that as our
+                # result.
+                if math.isnan(result) and allow_nan:
+                    pass
                 else:
-                    assert pos_clamper is not None
-                    clamped = pos_clamper(result)
-                if clamped != result and not (math.isnan(result) and allow_nan):
-                    self._cd.stop_example(discard=True)
-                    self._cd.start_example(DRAW_FLOAT_LABEL)
-                    self._write_float(clamped)
-                    result = clamped
+                    # if we drew a nan, and we don't allow nans, clamp it to
+                    # inf/ninf (depending on sign). Then clamp it further
+                    # based on our clampers.
+                    # This is to be explicit about weird nan interactions with
+                    # our clampers.
+                    if math.isnan(result) and not allow_nan:
+                        result = math.inf * math.copysign(1.0, result)
```

Comment:
1. We only create one of `neg_clamper` or `pos_clamper`, depending on bounds.
2. `self._draw_float(...)` can generate nans.
3. nans can be positive or negative, and in particular this sign can be "outside" of the specified bounds.

This combination of factors means that we may (1) generate a nan and (2) try to clamp it with a clamper that doesn't exist. We were already silently relying on the clamper to clamp nans to infs; I've made that behavior explicit here by not using clampers when a nan is involved. I think this is more clear to follow, but happy to hear other suggestions. (I did consider "always create both clampers" as a potential solution.)